src/mkfs.fat.c

```diff
@@ -1422,8 +1472,14 @@ static void write_tables(void)
 	for (y = 0; y < blank_fat_length; y++)
 	    writebuf(blank_sector, sector_size, "FAT");
     }
-    /* Write the root directory directly after the last FAT. This is the root
-     * dir area on FAT12/16, and the first cluster on FAT32. */
+    /* Write the root directory. On FAT12/16 it is directly after the last
+     * FAT. On FAT32 seek to root cluster. */
+    if (size_fat == 32) {
+	unsigned int cluster = le32toh(bs.fat32.root_cluster);
+	off_t sector = (off_t)start_data_sector +
+	    (cluster - 2) * bs.cluster_size * (sector_size / HARD_SECTOR_SIZE);
+	seekto(sector * HARD_SECTOR_SIZE, "root sector");
+    }
```

Comment:
I must admit that this code is hard to understand, specially because of `start_data_sector` variable which is in strange units. In PR #182 there is a proposal to cleanup of it and in code do not use strange units (fixed 512 or 1024).

So what about avoiding usage of `start_data_sector` and write it for example as this?

```suggestion
    if (size_fat == 32) {
	unsigned int root_cluster = le32toh(bs.fat32.root_cluster);
	off_t root_sector = reserved_sectors + nr_fats * fat_length + (root_cluster - 2) * bs.cluster_size;
	seekto(root_sector * sector_size, "root sector");
    }
```

The best would be if `start_data_sector` is in sector_size unit...

Reply:
I have adjusted the commit and also added an off_t cast to avoid integer overflow on large filesystems.